cmd/docker/docker.go

```diff
@@ -178,12 +178,24 @@ func setValidateArgs(dockerCli command.Cli, cmd *cobra.Command) {
 	})
 }
 
-func tryPluginRun(dockerCli command.Cli, cmd *cobra.Command, subcommand string) error {
+func tryPluginRun(dockerCli command.Cli, cmd *cobra.Command, orargs []string, subcommand string, useBuilderAlias bool) error {
 	plugincmd, err := pluginmanager.PluginRunCommand(dockerCli, subcommand, cmd)
 	if err != nil {
 		return err
 	}
 
+	// If build subcommand has changed to use buildx, user would expect
+	// "docker build" to always create a local docker image (default context
+	// builder). This is for better backward compatibility in case where a user
+	// could switch to a docker container builder with "docker buildx --use foo"
+	// which does not --load by default. Also makes sure that an arbitrary
+	// builder name is not being set in the command line or in the environment
+	// before setting the default context and keep "buildx install" behavior
+	// if being set (builder alias).
+	if orargs[0] == "build" && orargs[0] != subcommand && !useBuilderAlias && !hasBuilderName(plugincmd.Args, plugincmd.Env) {
```

Comment:
Can you explain what this is doing?
Maybe it would help to split that out into a function call with a test for various inputs.

Reply:
Refactored logic and added more tests. Let me know if that SGTY.